### PR TITLE
Support for Ghost Fields

### PR DESCRIPTION
**Pull request by ArquintL**
This PR depends on PRs #747 & #755 and, thus, targets `ghost-field-target-branch`, which corresponds to the branch used in #755.

This PR adds ghost fields, i.e., allows the use of `ghost` in front of field declarations. Note that struct embeddings can (so far) not be ghost. While ghost fields work mostly analogously to actual, non-ghost, fields, there are a few restrictions / differences to point out:
- A struct with ghost fields cannot implement an interface, however, a pointer to such a struct can (see `ghost-field-interface-simple01.gobra` and `ghost-field-interface-fail01.gobra`)
- Actual equality (`==`) on structs considers only actual fields, i.e., structs are equal modulo ghost fields. To include ghost fields in the comparison, ghost equality (`===`) can be used instead (see `ghost-field-comparison-simple01.gobra`)
- As a consequence, a Go map with a struct with ghost fields as a key type becomes challenging as two ghost-unequal key values can be actual-equal and, thus, map to the same entry at runtime. Although our encoding could be extended, we currently disallow such map types (see `ghost-field-map-fail01.gobra`)

### CODE REVIEW

**Comment by jcp19** [comment]
I will soon test this out in SCION to make sure that this works with the SCION idioms

**Comment by jcp19** [comment]
I just adapted the SCION code to be accepted by this new PR (https://github.com/viperproject/VerifiedSCION/pull/337). As such, I do NOT have any objections against merging this 🚀. Note that this comment is not a review, I did not look at the code.

File: src/main/scala/viper/gobra/frontend/Config.scala [hunk 1]
```diff
@@ -89,6 +89,31 @@ object MCE {
...
+
+  def merge(m1: Mode, m2: Mode): Mode = {
+    (m1, m2) match  {
+      case (All, _) | (_, All) => All
+      case (Impure, _) | (_, Impure) => Impure
+      case _ => Disabled
+    }
+  }
+}
+
```
Comment (jcp19) [inline src/main/scala/viper/gobra/frontend/Config.scala:116]: For some reason, Github is showing the diff for old commits that were merged in master already. Maybe we need to merge master here

Comment (ArquintL) [reply src/main/scala/viper/gobra/frontend/Config.scala]: I fixed it now

File: src/main/scala/viper/gobra/frontend/info/base/Type.scala [hunk 2]
```diff
@@ -99,15 +99,29 @@ object Type {
     case object Send extends ChannelModus("chan<-")
   }
 
-  case class StructT(clauses: ListMap[String, (Boolean, Type)], decl: PStructType, context: ExternalTypeInfo) extends ContextualType {
-    lazy val fieldsAndEmbedded: ListMap[String, Type] = clauses.map(removeFieldIndicator)
-    lazy val fields: ListMap[String, Type] = clauses.filter(isField).map(removeFieldIndicator)
-    lazy val embedded: ListMap[String, Type] = clauses.filterNot(isField).map(removeFieldIndicator)
-    private def isField(clause: (String, (Boolean, Type))): Boolean = clause._2._1
-    private def removeFieldIndicator(clause: (String, (Boolean, Type))): (String, Type) = (clause._1, clause._2._2)
+  trait StructClauseT extends Type {
```
Comment (jcp19) [inline src/main/scala/viper/gobra/frontend/info/base/Type.scala:102]: why are we making this a subtype of `Type`? We are adding a case that we have to dismiss every time we pattern match on `Type`, which sounds unnecessary.

File: src/main/scala/viper/gobra/frontend/info/implementation/property/Implements.scala [hunk 3]
```diff
@@ -89,9 +89,10 @@ trait Implements { this: TypeInfoImpl =>
     case _ => failedProp(s"$r is not an interface")
   }
 
-  /** Returns true if the type is supported for interfaces. All finite types are supported. */
+  /** Returns true if the type is supported for interfaces. All finite types are supported except for structs with ghost fields. */
   def supportedSortForInterfaces(t: Type): PropertyResult = {
-    failedProp(s"The type $t is not supported for interface", !isIdentityPreservingType(t))
+    failedProp(s"The type $t is not supported for interface", !isIdentityPreservingType(t)) and
+    failedProp(s"Structs containing ghost fields are not supported for interface", isStructTypeWithGhostFields(t))
   }
```
Comment (jcp19) [inline src/main/scala/viper/gobra/frontend/info/implementation/property/Implements.scala:96]: what is the intuition behind this restriction?

File: src/main/scala/viper/gobra/frontend/info/implementation/typing/TypeTyping.scala [hunk 4]
```diff
@@ -75,6 +77,19 @@ trait TypeTyping extends BaseTyping { this: TypeInfoImpl =>
     case t: PExpressionAndType => wellDefExprAndType(t).out ++ isType(t).out
   }
 
+  def isStructTypeWithGhostFields(t: Type): Boolean = t match {
+    case Single(st) => underlyingType(st) match {
+      case t: StructT =>
+        structMemberSet(t).collect {
+          case (_, f: Field) => f.ghost
+          case (_, e: Embbed) => isStructTypeWithGhostFields(e.context.typ(e.decl.typ))
```
Comment (jcp19) [inline src/main/scala/viper/gobra/frontend/info/implementation/typing/TypeTyping.scala:85]: shouldn't this yield true if `e` is a ghost embbed?

File: src/main/scala/viper/gobra/frontend/Desugar.scala [hunk 5]
```diff
@@ -4025,36 +4025,36 @@ object Desugar extends LazyLogging {
...
-        case (name, (false, typ)) => embeddedDeclD((name, typ), Addressability.field(addrMod), struct)(src)
+        case (name, t: StructFieldT) => fieldDeclD((name, t), Addressability.field(addrMod), struct)(src)
+        case (name, t: StructEmbeddedT) => embeddedDeclD((name, t), Addressability.field(addrMod), struct)(src)
       }.toVector
 
     def structMemberD(m: st.StructMember, addrMod: Addressability)(src: Meta): in.Field = m match {
-      case st.Field(decl, _, context) => fieldDeclD(decl, addrMod, context)(src)
-      case st.Embbed(decl, _, context) => embeddedDeclD(decl, addrMod, context)(src)
+      case st.Field(decl, isGhost, context) => fieldDeclD(decl, addrMod, isGhost, context)(src)
+      case st.Embbed(decl, isGhost, context) => embeddedDeclD(decl, addrMod, isGhost, context)(src)
```
[the remark this answers is not in the document]
Comment (ArquintL) [reply src/main/scala/viper/gobra/frontend/Desugar.scala:4034]: I'm adding two tests

**Review by jcp19 — APPROVED** [review]
LGTM! I would be happy to merge this when the tests go through

File: src/test/resources/regressions/features/ghost_field/ghost-field-comparison-simple01.gobra [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+// http://creativecommons.org/publicdomain/zero/1.0/
+
+// tests that Gobra correctly excludes ghost fields when comparing structs using the standard Go comparison
+
+package GhostFieldComparisonSimple01
+
+type Test struct {
+    actualField int
+    ghost ghostField int
+}
```
Comment (Felalolf) [inline src/test/resources/regressions/features/ghost_field/ghost-field-comparison-simple01.gobra:11]: Please also check fields that are structs with ghost fields.

**Review by Felalolf — COMMENTED** [review]
I think I found an issue. Please look at that. otherwise, it should be fine. I did not check all tests yet.

File: src/main/scala/viper/gobra/frontend/info/implementation/property/Implements.scala [hunk 7]
```diff
@@ -104,7 +110,7 @@ trait Implements { this: TypeInfoImpl =>
         case Type.NilType | Type.BooleanT | _: Type.IntT | Type.StringT => true
         case ut: Type.PointerT => go(ut.elem)
         case ut: Type.StructT =>
-          ut.clauses.forall{ case (_, (_, fieldType)) => go(fieldType) }
+          ut.clauses.forall{ case (_, info) => go(info.typ) }
```
Comment (Felalolf) [inline src/main/scala/viper/gobra/frontend/info/implementation/property/Implements.scala:113]: structs with ghost fields are not identity preserving. This also makes line 100 redundant.

**Review by Felalolf — COMMENTED** [review]
missed something

File: src/main/scala/viper/gobra/frontend/info/implementation/typing/ghost/separation/GhostTyping.scala [hunk 8]
```diff
@@ -140,8 +141,12 @@ trait GhostTyping extends GhostClassifier { this: TypeInfoImpl =>
       case e: PDot => resolve(e) match {
         case Some(s: ap.FieldSelection) =>
           val isGhostField = ghostIdClassification(s.id)
-          (underlyingType(typ(s.base)), isGhostField) match {
-            case (_, true) => isGhost // ghost fields are always ghost memory
+          val isGhostEmbedding = s.path.exists {
+            case MemberPath.Next(decl) => decl.ghost
```
[the remark this answers is not in the document]
Comment (ArquintL) [reply src/main/scala/viper/gobra/frontend/info/implementation/typing/ghost/separation/GhostTyping.scala:145]: I don't quite understand. I'm iterating here over the entire field selection's path

File: src/test/resources/regressions/features/ghost_field/ghost-field-comparison-simple01.gobra (hunk 6 above)
Comment (ArquintL) [reply src/test/resources/regressions/features/ghost_field/ghost-field-comparison-simple01.gobra:11]: I've added a corresponding testcase and actually uncovered & fixed an issue

**Review by Felalolf — APPROVED** [review]

**Pull request merged**